File: sdk/src/main/java/com/libreshockwave/util/AudioCodecUtils.java
package com.libreshockwave.util;

/**
 * Shared audio codec detection utilities used by SoundChunk and MediaChunk.
 */
public class AudioCodecUtils {

    private AudioCodecUtils() {}

    /**
     * Search for an MP3 sync frame within the given byte array.
     * Scans up to {@code searchLimit} bytes for a valid MP3 frame header
     * (sync word 0xFFE0+, valid version/layer/bitrate/sampleRate fields).
     *
     * @param data the audio data bytes
     * @param searchLimit maximum number of bytes to scan (typically 512)
     * @return true if a valid MP3 frame header was found
     */
    public static boolean containsMp3SyncFrame(byte[] data, int searchLimit) {
        if (data == null || data.length < 4) {
            return false;
        }
        int limit = Math.min(data.length - 4, searchLimit);
        for (int i = 0; i < limit; i++) {
            if ((data[i] & 0xFF) == 0xFF && (data[i + 1] & 0xE0) == 0xE0) {
                int version = (data[i + 1] >> 3) & 3;
                int layer = (data[i + 1] >> 1) & 3;
                int bitrateIdx = (data[i + 2] >> 4) & 0xF;
                int sampleRateIdx = (data[i + 2] >> 2) & 3;

                if (version != 1 && layer != 0 && bitrateIdx != 0 && bitrateIdx != 15 && sampleRateIdx != 3) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: cpp/src/util/AudioCodecUtils.cpp
#include "libreshockwave/util/AudioCodecUtils.hpp"

#include <algorithm>

namespace libreshockwave::util {

bool containsMp3SyncFrame(const std::vector<std::uint8_t>& data, int searchLimit) {
    if (data.size() < 4) {
        return false;
    }

    const int limit = std::min(static_cast<int>(data.size()) - 4, searchLimit);
    for (int index = 0; index < limit; ++index) {
        if (data[static_cast<std::size_t>(index)] == 0xFF &&
            (data[static_cast<std::size_t>(index + 1)] & 0xE0U) == 0xE0U) {
            const int version = (data[static_cast<std::size_t>(index + 1)] >> 3) & 3;
            const int layer = (data[static_cast<std::size_t>(index + 1)] >> 1) & 3;
            const int bitrateIndex = (data[static_cast<std::size_t>(index + 2)] >> 4) & 0xF;
            const int sampleRateIndex = (data[static_cast<std::size_t>(index + 2)] >> 2) & 3;

            if (version != 1 && layer != 0 && bitrateIndex != 0 && bitrateIndex != 15 && sampleRateIndex != 3) {
                return true;
            }
        }
    }
    return false;
}

} // namespace libreshockwave::util
